### DISCUSSION TITLE
Feasibility of custom shading

### DISCUSSION THREAD

**Discussion by mikael-wants-to-visit-disneyland** [gfx-rs/wgpu#3257] [General]
Is it possible to achieve a hatching effect? I want to be able to control the hatch line density and thickness based on the surface's normal, and if possible, whether there is a shadow on it. However, I cannot find out how to paint a dynamic texture on a plane, such that it is free from perspective.

![Frame 1 (9)](https://user-images.githubusercontent.com/8318416/205433925-0ccc9f5e-0099-43d8-aaa7-b9de63f02525.png)

**Comment by kpreid** [comment]
Effects like this hatching can be done with reference to screen-space coordinates by using an `@builtin(position)` input to the fragment shader. Use that as texture coordinates or procedurally compute your lines — either way it will keep the hatching “free from perspective” and with a scale referenced to the viewport rather than the vertices' world positions.

**Reply by mikael-wants-to-visit-disneyland** [reply]
Makes sense. Thank you.
